### PR TITLE
improved README, added time-series subsetting example

### PR DESCRIPTION
**Pull request by lsilvest**
Hopefully this is complete enough to finalize this package. We can then see how things go and which directions to take for further development.

### CODE REVIEW

**Comment by eddelbuettel** [comment]
How would you feel about moving `nanotime` from `Depends:` to `Imports:`?  It needs one `library(nanotime)` in the unit test file but is otherwise 'more modern'.

**Comment by eddelbuettel** [comment]
Also made the repo 'public' as the links to the issue tickets in `DESCRIPTION` and `README.md` need it.

**Comment by lsilvest** [comment]
> How would you feel about moving `nanotime` from `Depends:` to `Imports:`?  It needs one `library(nanotime)` in the unit test file but is otherwise 'more modern'.  

Yes, that seems good.

**Comment by eddelbuettel** [comment]
So I think it's ready to ship. I get

```sh
edd@rob:~/git/dtts(readme_changes)$ rdcc.sh dtts_0.1.0.tar.gz           # wrapper for R-devel CMD check --as-cran                                  
* using log directory ‘/tmp/dtts.Rcheck’           
* using R Under development (unstable) (2022-03-04 r81849)
* using platform: x86_64-pc-linux-gnu (64-bit)
* using session charset: UTF-8                                                                           
* using option ‘--as-cran’                     
* checking for file ‘dtts/DESCRIPTION’ ... OK                                                            
* checking extension type ... Package      
* this is package ‘dtts’ version ‘0.1.0’       
* package encoding: UTF-8                        
* checking CRAN incoming feasibility ... NOTE                                                            
Maintainer: ‘Dirk Eddelbuettel <edd@debian.org>’
                                                                                                         
New submission                                     
* checking package namespace information ... OK
* checking package dependencies ... OK
* checking if this is a source package ... OK
* checking if there is a namespace ... OK                                                                
* checking for executable files ... OK
* checking for hidden files and directories ... OK
* checking for portable file names ... OK
* checking for sufficient/correct file permissions ... OK
* checking serialization versions ... OK                                                                 
* checking whether package ‘dtts’ can be installed ... OK
* checking installed package size ... OK
* checking package directory ... OK
* checking for future file timestamps ... OK
* checking DESCRIPTION meta-information ... OK
* checking top-level files ... OK
* checking for left-over files ... OK
* checking index information ... OK
* checking package subdirectories ... OK
* checking R files for non-ASCII characters ... OK
* checking package subdirectories ... OK
* checking R files for non-ASCII characters ... OK
* checking R files for syntax errors ... OK
* checking whether the package can be loaded ... OK
* checking whether the package can be loaded with stated dependencies ... OK
* checking whether the package can be unloaded cleanly ... OK
* checking whether the namespace can be loaded with stated dependencies ... OK
* checking whether the namespace can be unloaded cleanly ... OK
* checking use of S3 registration ... OK
* checking dependencies in R code ... OK
* checking S3 generic/method consistency ... OK
* checking replacement functions ... OK
* checking foreign function calls ... OK
* checking R code for possible problems ... OK
* checking Rd files ... OK
* checking Rd metadata ... OK
* checking Rd line widths ... OK
* checking Rd cross-references ... OK
* checking for missing documentation entries ... OK
* checking for code/documentation mismatches ... OK
* checking Rd \usage sections ... OK
* checking Rd contents ... OK
* checking for unstated dependencies in examples ... OK
* checking line endings in shell scripts ... OK
* checking line endings in C/C++/Fortran sources/headers ... OK
* checking line endings in Makefiles ... OK
* checking compilation flags in Makevars ... OK
* checking for GNU extensions in Makefiles ... OK
* checking for portable use of $(BLAS_LIBS) and $(LAPACK_LIBS) ... OK
* checking use of PKG_*FLAGS in Makefiles ... OK
* checking use of SHLIB_OPENMP_*FLAGS in Makefiles ... OK
* checking pragmas in C/C++ headers and code ... OK
* checking compilation flags used ... OK
* checking compiled code ... OK
* checking examples ... OK
* checking for unstated dependencies in ‘tests’ ... OK
* checking tests ...
  Running ‘tinytest.R’
 OK
* checking PDF version of manual ... OK
* checking for non-standard things in the check directory ... OK
* checking for detritus in the temp directory ... OK
* DONE

Status: 1 NOTE
See
  ‘/tmp/dtts.Rcheck/00check.log’
for details.


edd@rob:~/git/dtts(readme_changes)$ 
```

**Comment by lsilvest** [comment]
Yes, super cool!

**Comment by eddelbuettel** [comment]
Sadly windows is not our friend (yet).  We fail on r-release and r-devel:

```
Dear package maintainer,

this notification has been generated automatically.
Your package dtts_0.1.0.tar.gz has been built (if working) and checked for Windows.
Please check the log files and (if working) the binary package at:
https://win-builder.r-project.org/DSQXWH9Vt6vB
The files will be removed after roughly 72 hours.
Installation time in seconds: 9 
Check time in seconds: 12
Status: 1 ERROR, 1 NOTE
R Under development (unstable) (2022-03-04 r81849 ucrt)
```

and

```
Dear package maintainer,

this notification has been generated automatically.
Your package dtts_0.1.0.tar.gz has been built (if working) and checked for Windows.
Please check the log files and (if working) the binary package at:
https://win-builder.r-project.org/mVtcSU21Z0zp
The files will be removed after roughly 72 hours.
Installation time in seconds: 18
Check time in seconds: 22
Status: 1 ERROR, 1 NOTE
R version 4.1.2 (2021-11-01
```

Compilation fails over `localtime`:  

```
* installing *source* package 'dtts' ...
** using staged installation
** libs
g++  -std=gnu++11 -I"D:/RCompile/recent/R/include" -DNDEBUG  -I'D:/RCompile/CRANpkg/lib/4.2/Rcpp/include' -I'D:/RCompile/CRANpkg/lib/4.2/RcppCCTZ/include' -I'D:/RCompile/CRANpkg/lib/4.2/RcppDate/include' -I'D:/RCompile/CRANpkg/lib/4.2/nanotime/include'   -I"d:/rtools42/x86_64-w64-mingw32.static.posix/include"     -O2 -Wall  -mfpmath=sse -msse2 -mstackrealign  -c RcppExports.cpp -o RcppExports.o
g++  -std=gnu++11 -I"D:/RCompile/recent/R/include" -DNDEBUG  -I'D:/RCompile/CRANpkg/lib/4.2/Rcpp/include' -I'D:/RCompile/CRANpkg/lib/4.2/RcppCCTZ/include' -I'D:/RCompile/CRANpkg/lib/4.2/RcppDate/include' -I'D:/RCompile/CRANpkg/lib/4.2/nanotime/include'   -I"d:/rtools42/x86_64-w64-mingw32.static.posix/include"     -O2 -Wall  -mfpmath=sse -msse2 -mstackrealign  -c align.cpp -o align.o
align.cpp: In function 'std::ostream& operator<<(std::ostream&, const std::chrono::time_point<_Clock, _Duration1>&)':
align.cpp:29:3: error: there are no arguments to 'localtime_r' that depend on a template parameter, so a declaration of 'localtime_r' must be available [-fpermissive]
   29 |   localtime_r(&time, &tm);
      |   ^~~~~~~~~~~
align.cpp:29:3: note: (if you use '-fpermissive', G++ will accept your code, but allowing the use of an undeclared name is deprecated)
make: *** [D:/RCompile/recent/R/etc/x64/Makeconf:259: align.o] Error 1
ERROR: compilation failed for package 'dtts'
* removing 'd:/RCompile/CRANguest/R-devel/lib/dtts'
* ```

**Comment by eddelbuettel** [comment]
I dropped a quick test for `_WIN32` around. Else maybe this is promising:  
https://stackoverflow.com/questions/38034033/c-localtime-this-function-or-variable-may-be-unsafe/38034148#38034148

**Comment by eddelbuettel** [comment]
Looks like this did it:  https://stackoverflow.com/a/38034148/143305

Current diff:

```diff
@@ -22,11 +22,15 @@ template<typename Clock, typename Duration>
 std::ostream &operator<<(std::ostream &stream,
                          const std::chrono::time_point<Clock, Duration> &time_point) {
   const time_t time = Clock::to_time_t(time_point);
-#if __GNUC__ > 4 || \
-    ((__GNUC__ == 4) && __GNUC_MINOR__ > 8 && __GNUC_REVISION__ > 1)
+#if (__GNUC__ > 4 || ((__GNUC__ == 4) && __GNUC_MINOR__ > 8 && __GNUC_REVISION__ > 1))
   // Maybe the put_time will be implemented later?
   struct tm tm;
+  // thanks to https://stackoverflow.com/a/38034148/143305 for the next test
+  #if defined(__unix__)
   localtime_r(&time, &tm);
+  #elif defined(_MSC_VER)
+  localtime_s(&tm, &time);
+  #endif
   return stream << std::put_time(&tm, "%c"); // Print standard date&time
 #else
   char buffer[26];

```

Passed here:  https://builder.r-hub.io/status/dtts_0.1.0.tar.gz-0415efbf471a4eaca733bddfe975d122

**Comment by eddelbuettel** [comment]
And good at win-builder too:

#### r-release

```
this notification has been generated automatically.
Your package dtts_0.1.0.tar.gz has been built (if working) and checked for Windows.
Please check the log files and (if working) the binary package at:
https://win-builder.r-project.org/o8tDQBwHrJKG
The files will be removed after roughly 72 hours.
Installation time in seconds: 60
Check time in seconds: 136
Status: 1 NOTE
R version 4.1.2 (2021-11-01)
```

#### r-devel

```
this notification has been generated automatically.
Your package dtts_0.1.0.tar.gz has been built (if working) and checked for Windows.
Please check the log files and (if working) the binary package at:
https://win-builder.r-project.org/JNS8A6tFLSBj
The files will be removed after roughly 72 hours.
Installation time in seconds: 17
Check time in seconds: 68 
Status: 1 NOTE
R Under development (unstable) (2022-03-04 r81849 ucrt)
```

The note is simply that we're new:

```
* using log directory 'd:/RCompile/CRANguest/R-release/dtts.Rcheck'
* using R version 4.1.2 (2021-11-01)
* using platform: x86_64-w64-mingw32 (64-bit)
* using session charset: ISO8859-1
* checking for file 'dtts/DESCRIPTION' ... OK
* checking extension type ... Package
* this is package 'dtts' version '0.1.0'
* package encoding: UTF-8
* checking CRAN incoming feasibility ... NOTE
Maintainer: 'Dirk Eddelbuettel <edd@debian.org>'

New submission
* checking package namespace information ... OK
* checking package dependencies ... OK
* checking if this is a source package ... OK
* checking if there is a namespace ... OK
* checking for hidden files and directories ... OK
* checking for portable file names ... OK
* checking serialization versions ... OK
* checking whether package 'dtts' can be installed ... OK
* checking installed package size ... OK
* checking package directory ... OK
* checking for future file timestamps ... OK
* checking DESCRIPTION meta-information ... OK
* checking top-level files ... OK
* checking for left-over files ... OK
* checking index information ... OK
* checking package subdirectories ... OK
* checking R files for non-ASCII characters ... OK
* checking R files for syntax errors ... OK
* loading checks for arch 'i386'
** checking whether the package can be loaded ... OK
** checking whether the package can be loaded with stated dependencies ... OK
** checking whether the package can be unloaded cleanly ... OK
** checking whether the namespace can be loaded with stated dependencies ... OK
** checking whether the namespace can be unloaded cleanly ... OK
** checking loading without being on the library search path ... OK
** checking use of S3 registration ... OK
* loading checks for arch 'x64'
** checking whether the package can be loaded ... OK
** checking whether the package can be loaded with stated dependencies ... OK
** checking whether the package can be unloaded cleanly ... OK
** checking whether the namespace can be loaded with stated dependencies ... OK
** checking whether the namespace can be unloaded cleanly ... OK
** checking loading without being on the library search path ... OK
** checking use of S3 registration ... OK
* checking dependencies in R code ... OK
* checking S3 generic/method consistency ... OK
* checking replacement functions ... OK
* checking foreign function calls ... OK
* checking R code for possible problems ... [12s] OK
* checking Rd files ... [2s] OK
* checking Rd metadata ... OK
* checking Rd line widths ... OK
* checking Rd cross-references ... OK
* checking for missing documentation entries ... OK
* checking for code/documentation mismatches ... OK
* checking Rd \usage sections ... OK
* checking Rd contents ... OK
* checking for unstated dependencies in examples ... OK
* checking line endings in shell scripts ... OK
* checking line endings in C/C++/Fortran sources/headers ... OK
* checking line endings in Makefiles ... OK
* checking compilation flags in Makevars ... OK
* checking for GNU extensions in Makefiles ... OK
* checking for portable use of $(BLAS_LIBS) and $(LAPACK_LIBS) ... OK
* checking use of PKG_*FLAGS in Makefiles ... OK
* checking use of SHLIB_OPENMP_*FLAGS in Makefiles ... OK
* checking pragmas in C/C++ headers and code ... OK
* checking compiled code ... OK
* checking examples ...
** running examples for arch 'i386' ... [2s] OK
** running examples for arch 'x64' ... [3s] OK
* checking for unstated dependencies in 'tests' ... OK
* checking tests ...
** running tests for arch 'i386' ... [4s] OK
  Running 'tinytest.R' [3s]
** running tests for arch 'x64' ... [4s] OK
  Running 'tinytest.R' [4s]
* checking PDF version of manual ... OK
* checking for detritus in the temp directory ... OK
* DONE
Status: 1 NOTE
```

So :shipit:  (<-- that's an emoji called 'shipit' and I have no idea where the joke comes from but I play along...)

**Comment by eddelbuettel** [comment]
It has been shipped.  Now up to ten working days to hear about it...

```
Dear maintainer,

package dtts_0.1.0.tar.gz has been auto-processed and is pending a manual inspection of this new CRAN submission. A CRAN team member will typically respond to you within the next 10 working days. For technical
reasons you may receive a second copy of this message when a team member triggers a new check.

Log dir: <https://win-builder.r-project.org/incoming_pretest/dtts_0.1.0_20220307_025617/>
The files will be removed after roughly 7 days.
Installation time in seconds: 18
Check time in seconds: 68
R Under development (unstable) (2022-03-04 r81849 ucrt)

Pretests results:
Windows: <https://win-builder.r-project.org/incoming_pretest/dtts_0.1.0_20220307_025617/Windows/00check.log>
Status: 1 NOTE
Debian: <https://win-builder.r-project.org/incoming_pretest/dtts_0.1.0_20220307_025617/Debian/00check.log>
Status: 1 NOTE



No strong reverse dependencies to be checked.

Best regards,
CRAN teams' auto-check service

[-- Attachment #2: "00details.log" --]                                                                                                                                                                             
[-- Type: text/plain, Encoding: 7bit, Size: 0.2K --]                                                                                                                                                               

Flavor: r-devel-linux-x86_64-debian-gcc, r-devel-windows-x86_64
Check: CRAN incoming feasibility, Result: NOTE  
  Maintainer: 'Dirk Eddelbuettel <edd@debian.org>'

  New submission   

```

**Comment by lsilvest** [comment]
Wow, that is fantastic and really exciting!

**Comment by eddelbuettel** [comment]
Thanks for all your work on this, and on `nanotime`!

**Comment by lsilvest** [comment]
Thanks for all your work!!

File: README.md [hunk 1]
```diff
@@ -253,11 +251,27 @@ Which produces:
...
+### Time-series subsetting
+
+Using `nanoival`, it is possible to do complex subsetting of a time-series:
+
+~~~ R
+one_second <- 1e9
+index <- seq(nanotime("2022-12-12 12:12:10+00:00"), length.out=10, by=one_second)
+dts <- data.table(index=index, data=1:length(index), key="index")
+ival <- c(as.nanoival("-2022-12-12 12:12:10+00:00 -> 2022-12-12 12:12:14+00:00-"),
+          as.nanoival("+2022-12-12 12:12:18+00:00 -> 2022-12-12 12:12:20+00:00+"))
```
Comment (jangorecki) [inline README.md:264]: Would probably look more tidy :D by writing `as.nanoival` once
```r
as.nanoival(c(
 ...
))
```

**Comment by eddelbuettel** [comment]
Woot, woot, this worked well and fairly quickly:

image

So merging, and welcome to release 0.1.0.

File: README.md (hunk 1 above)
Comment (lsilvest) [reply README.md:264]: Indeed, thanks!

Comment (eddelbuettel) [reply README.md:264]: We'll fold this into 0.1.1 which we'll do in a few days to add CRAN and CI badges to the README.md and other small things now that we are on CRAN.

**Pull request merged**